src/Http/Http.Extensions/src/RequestDelegateFactory.cs

```diff
@@ -643,12 +677,112 @@ private static Expression AddResponseWritingToMethodCall(Expression methodCall,
             }
         }
 
-        private static Expression GetValueFromProperty(Expression sourceExpression, string key)
+        private static Func<object?, HttpContext, Task> HandleRequestBodyAndCompileRequestDelegateForForm(
+            Expression responseWritingMethodCall,
+            FactoryContext factoryContext)
+        {
+            Debug.Assert(factoryContext.FormRequestBodyParameter is not null, "factoryContext.FormRequestBodyParameter is null for a form body.");
+
+            var parameterTypeName = TypeNameHelper.GetTypeDisplayName(factoryContext.FormRequestBodyParameter.ParameterType, fullName: false);
+            var parameterName = factoryContext.FormRequestBodyParameter.Name;
+
+            Debug.Assert(parameterName is not null, "CreateArgument() should throw if parameter.Name is null.");
+
+            if (factoryContext.ParameterBinders.Count > 0)
+            {
+                // We need to generate the code for reading from the body or form before calling into the delegate
+                var continuation = Expression.Lambda<Func<object?, HttpContext, object?, object?[], Task>>(
+                responseWritingMethodCall, TargetExpr, HttpContextExpr, BodyValueExpr, BoundValuesArrayExpr).Compile();
+
+                // Looping over arrays is faster
+                var binders = factoryContext.ParameterBinders.ToArray();
+                var count = binders.Length;
+
+                return async (target, httpContext) =>
+                {
+                    // Run these first so that they can potentially read and rewind the body
+                    var boundValues = new object?[count];
+
+                    for (var i = 0; i < count; i++)
+                    {
+                        boundValues[i] = await binders[i](httpContext);
+                    }
+
+                    object? formValue = null;
+                    var feature = httpContext.Features.Get<IHttpRequestBodyDetectionFeature>();
+                    if (feature?.CanHaveBody == true)
+                    {
+                        if (!httpContext.Request.HasJsonContentType())
+                        {
+                            Log.UnexpectedFormContentType(httpContext, httpContext.Request.ContentType, factoryContext.ThrowOnBadRequest);
+                            httpContext.Response.StatusCode = StatusCodes.Status415UnsupportedMediaType;
+                            return;
+                        }
+                        try
+                        {
+                            formValue = await httpContext.Request.ReadFormAsync();
```

Comment:
Because we don't have an anti-forgery story for minimal APIs yet, we want to throw a `BadHttpRequestException` if there's a "Cookie" or "Authorization" request header. Or if there's a client certificate. @blowdart 

I'll file a follow up issue to add anti-forgery support when we merge this PR.